src/anet.c

```diff
@@ -620,3 +620,53 @@ int anetFormatFdAddr(int fd, char *buf, size_t buf_len, int fd_to_str_type) {
     anetFdToString(fd,ip,sizeof(ip),&port,fd_to_str_type);
     return anetFormatAddr(buf, buf_len, ip, port);
 }
+
+/* Create a pipe buffer with given flags for read end and write end. */
+int anetPipe(int fds[2], int read_flags, int write_flags) {
+    int pipe_flags = 0;
+#if defined(__linux__) || defined(__FreeBSD__)
+    /* When it leverages the pipe2() to create pipe,
+     * there is no harm to set O_CLOEXEC to prevent fd leaks. */
+    pipe_flags = O_CLOEXEC | (read_flags & write_flags);
+    if (pipe2(fds, pipe_flags)) {
+        if (errno != ENOSYS && errno != EINVAL)
+            return -1;
+        pipe_flags = 0;
+    } else {
+        if ((O_CLOEXEC | read_flags) == (O_CLOEXEC | write_flags))
+            return 0;
+        /* Clear the flags which have already been set using pipe2. */
+        read_flags &= ~pipe_flags;
+        write_flags &= ~pipe_flags;
+    }
+#endif
+
+    if (pipe_flags == 0 && pipe(fds))
+        return -1;
+
+    /* File descriptor flags.
+     * Currently, only one such flag is defined: FD_CLOEXEC, the close-on-exec flag. */
```

Comment:
thing brings me to another thought. are we sure we can mix these two flags in the same variable?
is there any guarantee that they don't collide? (considering they where intended to be used by different "functions")

Reply:
They don't collide. Pipe2 takes them in the same variable.

Reply:
> thing brings me to another thought. are we sure we can mix these two flags in the same variable? is there any guarantee that they don't collide? (considering they where intended to be used by different "functions")

Yes, there is such a guarantee because the only file descriptor flag: `FD_CLOEXEC/O_CLOEXEC` is now treated specially in the current `anetPipe()`, which makes sure this flag won't interfere with other file status flags like `O_NONBLOCK`.

Reply:
the argument about pipe2 taking them in the same variable applies only to Linux.. maybe in another system they can be overlapping?

Reply:
I guess we need to use `FD_CLOEXEC` when it's set with fcntl, e.g.

```diff
     if (read_flags & O_CLOEXEC)
-        if (fcntl(fds[0], F_SETFD, O_CLOEXEC))
+        if (fcntl(fds[0], F_SETFD, FD_CLOEXEC))
             goto error;
```